src/permission/permission.h

```diff
@@ -0,0 +1,86 @@
+#ifndef SRC_PERMISSION_PERMISSION_H_
+#define SRC_PERMISSION_PERMISSION_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "node_options.h"
+#include "permission/child_process_permission.h"
+#include "permission/fs_permission.h"
+#include "permission/permission_node.h"
+#include "permission/worker_permission.h"
+#include "v8.h"
+
+#include <iostream>
+#include <map>
+
+namespace node {
+
+class Environment;
+
+namespace permission {
+
+#define THROW_IF_INSUFFICIENT_PERMISSIONS(env, perm_, resource_, ...)          \
+  do {                                                                         \
+    if (UNLIKELY(!(env)->permission()->is_granted(perm_, resource_))) {        \
+      node::permission::Permission::ThrowAccessDenied((env), perm_);           \
+      return __VA_ARGS__;                                                      \
+    }                                                                          \
+  } while (0)
+
+class Permission {
+ public:
+  Permission() : enabled_(false) {
+    std::shared_ptr<PermissionNode> fs = std::make_shared<FSPermission>();
+    std::shared_ptr<PermissionNode> child_p =
+        std::make_shared<ChildProcessPermission>();
+    std::shared_ptr<PermissionNode> worker_t =
+        std::make_shared<WorkerPermission>();
+#define V(Name, _, __)                                                         \
+  nodes_.insert(std::make_pair(PermissionScope::k##Name, fs));
+    FILESYSTEM_PERMISSIONS(V)
+#undef V
+#define V(Name, _, __)                                                         \
+  nodes_.insert(std::make_pair(PermissionScope::k##Name, child_p));
+    CHILD_PROCESS_PERMISSIONS(V)
+#undef V
+#define V(Name, _, __)                                                         \
+  nodes_.insert(std::make_pair(PermissionScope::k##Name, worker_t));
+    WORKER_THREADS_PERMISSIONS(V)
+#undef V
+  }
+
+  inline bool is_granted(const PermissionScope permission, const char* res) {
+    if (!enabled_) return true;
+    auto perm_node = nodes_.find(permission);
+    if (perm_node != nodes_.end()) {
+      return perm_node->second->is_granted(permission, res);
+    }
+    return false;
```

Comment:
Maybe? But it would require creating an `is_granted_fs`, `is_granted_worker`, and so on. The same applies to `THROW_IF_...`. Do you have another approach in mind?

Reply:
I’m not sure I understand what you’re saying – this shouldn’t require creating any other functions? The only suggestion here is to put these lines into a separate function that doesn’t get inlined, it’s not really changing anything structurally.

Reply:
Sorry, I thought you meant to create separate functions for each permission to avoid the nodes_.find call. Fixed in e91a563879.